File: services/api-phoenix/test/app/compliance_test.exs
defmodule App.ComplianceTest do
  use ExUnit.Case

  # Read profiles.json straight from disk so we test the same catalog the
  # loader reads at runtime (App.Compliance reads from File.cwd!()/compliance).
  defp profiles do
    path = Path.join([File.cwd!(), "compliance", "profiles.json"])
    path |> File.read!() |> Jason.decode!() |> Map.fetch!("profiles")
  end

  test "every profile has the same control keys (uniform catalog)" do
    profs = profiles()
    refute profs == %{}

    [first | rest] = Map.values(profs)
    baseline_keys = first |> Map.fetch!("controls") |> Map.keys() |> MapSet.new()

    Enum.each(rest, fn entry ->
      keys = entry |> Map.fetch!("controls") |> Map.keys() |> MapSet.new()
      assert keys == baseline_keys
    end)
  end

  test "itsg-33 sets data_residency=canada and fips_crypto=true" do
    controls = profiles() |> Map.fetch!("itsg-33") |> Map.fetch!("controls")
    assert controls["data_residency"] == "canada"
    assert controls["fips_crypto"] == true
  end

  test "loader with COMPLIANCE_PROFILE=itsg-33 returns profile itsg-33" do
    prev = System.get_env("COMPLIANCE_PROFILE")
    System.put_env("COMPLIANCE_PROFILE", "itsg-33")

    try do
      App.Compliance.setup()
      active = App.Compliance.active()

      assert active.profile == "itsg-33"
      assert active.controls["data_residency"] == "canada"
      assert active.controls["fips_crypto"] == true
    after
      if prev, do: System.put_env("COMPLIANCE_PROFILE", prev), else: System.delete_env("COMPLIANCE_PROFILE")
    end
  end
end

File: services/protocol-ws-node/src/compliance.test.ts
import { describe, it, expect, beforeEach } from 'vitest'
import { readFileSync } from 'node:fs'
import { join } from 'node:path'
import { activeCompliance } from './compliance'

// Read the raw catalog directly so the test verifies the source of truth,
// not the loader's cache. The loader is exercised separately below.
const raw = JSON.parse(
  readFileSync(join(process.cwd(), 'compliance', 'profiles.json'), 'utf8'),
) as {
  device: string
  controlMeta: Record<string, unknown>
  profiles: Record<string, {
    name: string; priority: string; jurisdiction: string
    controls: Record<string, string | number | boolean>
  }>
}

describe('compliance catalog invariants', () => {
  it('every profile shares the exact same control keys', () => {
    const ids = Object.keys(raw.profiles)
    expect(ids.length).toBeGreaterThan(0)

    const reference = Object.keys(raw.profiles[ids[0]].controls).sort()
    expect(reference.length).toBeGreaterThan(0)

    for (const id of ids) {
      const keys = Object.keys(raw.profiles[id].controls).sort()
      expect(keys, `profile "${id}" control keys must match the uniform set`).toEqual(reference)
    }
  })

  it('control keys match the declared controlMeta keys', () => {
    const metaKeys = Object.keys(raw.controlMeta).sort()
    const ids = Object.keys(raw.profiles)
    const controlKeys = Object.keys(raw.profiles[ids[0]].controls).sort()
    expect(controlKeys).toEqual(metaKeys)
  })

  it('itsg-33 enforces Canadian data residency and FIPS crypto', () => {
    const itsg = raw.profiles['itsg-33']
    expect(itsg, 'itsg-33 profile must exist').toBeDefined()
    // This device's Canada-residency signal is the profile priority field.
    // There is no separate data_residency control key in the protocol catalog.
    expect(itsg.priority).toBe('canada')
    expect(itsg.controls.fips_crypto).toBe(true)
  })
})

describe('activeCompliance loader', () => {
  beforeEach(() => {
    delete process.env.COMPLIANCE_PROFILE
  })

  it('selecting itsg-33 returns the itsg-33 profile', () => {
    process.env.COMPLIANCE_PROFILE = 'itsg-33'
    const view = activeCompliance()
    expect(view.profile).toBe('itsg-33')
    expect(view.name).toBe(raw.profiles['itsg-33'].name)
    expect(view.jurisdiction).toBe(raw.profiles['itsg-33'].jurisdiction)
    expect(view.controls.fips_crypto).toBe(true)
  })

  it('defaults to baseline when COMPLIANCE_PROFILE is unset', () => {
    const view = activeCompliance()
    expect(view.profile).toBe('baseline')
    expect(view.name).toBe(raw.profiles['baseline'].name)
  })
})
